x-pack/plugins/alerting/server/authorization/alerting_authorization.ts

```diff
@@ -127,6 +125,12 @@ export class AlertingAuthorization {
         return new Set();
       });
 
+    this.spaceId = getSpace(request)
+      .then((maybeSpace) => maybeSpace?.id)
+      .catch((e) => {
+        return undefined;
```

Comment:
@dhurley14 I think we'd decided on returning `undefined` here, but wanted to open up discussion as to what the expected behavior is.

@gmmorris @ymao1 @legrego 
What should the expected behavior be here? We see that existing code in the constructor did not throw when `getSpace()` throws a 403 or error, it just sets the feature ids `Set` to an empty `Set`. Are there any consequences we're not thinking of when we set `spaceId` to `undefined`. Our integration tests are showing that we get what's expected (we still end up with a 403 being thrown down the line).

Reply:
In order to answer this I'd want to know all the reasons why `getSpace` might throw an error. Off the top of my head, that could include:

1) The user is not authorized to access the space
2) The space does not exist
3) A transient network error
4) Timeout due to an overloaded cluster
5) A bug in Kibana
6) Probably other reasons

So in all of these cases, `this.spaceId` would resolve to `undefined`, which means that the downstream code would potentially treat any of the above cases as if the `spaces` plugin was disabled. Without having a firm grasp on the big picture here, this feels error prone to me. I would rather us handle expected/unexpected errors explicitly, rather than relying on _something_ downstream to figure this out.

Reply:
So the reason we were seeing it throw in this case was because the user is not authorized to access the space. I agree that it seems weird to have something downstream catch this error, but it seems that's already what the existing code does. 

I definitely wouldn't want the downstream code to treat it as if spaces is disabled.

Reply:
> I agree that it seems weird to have something downstream catch this error, but it seems that's already what the existing code does.

Can we find exactly where we are throwing the `403` error? It would be good to make sure that there isn't anything happening before this that could accidentally mistake `undefined` to mean that the spaces plugin is disabled.

Reply:
Instead of setting `spaceId` in the constructor, could we modify the `getSpaceId()` function to dynamically get the space id? Whatever errors come from unauthorized users would then be handled by the calling function. That would prevent us from having to set some sort of state within this class to differentiate between spaces disabled and unprivileged user.

```
public async getSpaceId(): Promise<string | undefined> {
    return getSpace(request).then((maybeSpace) => maybeSpace?.id);
}
```

Reply:
The 403 comes from here:
https://github.com/elastic/kibana/blob/master/x-pack/plugins/security/server/spaces/secure_spaces_client_wrapper.ts#L161

And bubbles all the way up to:
https://github.com/elastic/kibana/blob/master/x-pack/plugins/alerting/server/authorization/alerting_authorization.ts#L105

and here:
https://github.com/elastic/kibana/blob/master/x-pack/plugins/alerting/server/authorization/alerting_authorization.ts#L124

Without the catch, the error is swallowed. In my code I add the catch.

Reply:
I'm in favor of @ymao1 's suggestion. Less state we store the better 👍

Reply:
Turns out...there's a non async function already to get the space id! So updated code to use this.